### ISSUE TITLE
Implement Flexbox solution to tooltips

### ISSUE THREAD

**Issue by a22moamy** [HGustavs/LenaSYS#15664]
## Background
We have had many issues with tooltips being placed weirdly. This is because they're all placed using pixel measurements, which results in them breaking whenever the form is updated somewhat. 

## The issue
To finally end this evil spiral and make a more long-lived solution, we have decided to switch to using Flexbox for all tooltips in DuggaSYS. With your fix, tooltips should be automatically aligned next to an input without having to cook up the perfect amount of pixels to add. 

Your solution should look good on wider screens (PC and laptops). Adjustments for mobile devices will be handled in another future issue.

**Comment by a22chrfa** [comment]
I ended up not using flexbox because of the extra layer of a wrapper div.

In it's original form the html structure looked like this:

`<div class="inputwrapper">
    <span> course code: </span>
    <div class="formdialog">
        <span> text </span>
    </div>
    <p> text </p>
    <input> inputfield </input>
</div>`

Instead, this is the solution:

`<div class="inputwrapper">
    <p> text </p>
    <div class="dialogwrapper">
        <div class="formdialog">
            <span> text </span>
        </div>
        <input> inputfield </input>
    </div>
</div>`

**Why another wrapper?**
Mostly semantics. 
I could've moved the input into the formdialog div and achieved the same thing with some editing. But it seems semantically incorrect to move an input field into a div called "formdialog". Secondly, the formdialog class has functions tied to it so I didn't think renaming and/or editing it too much would be a good idea.

With some reference to the previous comment. An absolutely positioned element will be anchored to it's closest ancestor that isn't static. By making the dialogwrapper have "position: relative" with the formdialog absolute, it will now be anchored to the dialogwrapper. Also, since formdialog is taken out of normal flow due to it's absolute position, the only element remaining inside the dialogwrapper is the input, making the dialogwrapper-div always be the same exact size and position as the input. When we position the formdialog now, it will effectively always be anchored to the input element.

With some adjusments to the css, the result looks like this:
![image](https://github.com/HGustavs/LenaSYS/assets/128893264/d7a227d5-f9f9-4694-a806-36e84c73b01a)

What needs to be done now is removing all the specific css/inline rules that are applied to all the other validation texts. Theoretically, once this is done. Whenever a valdiation text is added with the same html structure, the validation text should fit automatically. Hopefully nothing unexpected shows up.

**Comment by a20jonda** [comment]
The wiki should probably be updated after this is done.
https://github.com/HGustavs/LenaSYS/wiki/Popup-form-conventions#currently-known-popup-forms-on-lenasys

**Comment by a22chrfa** [comment]
Log:
Noticed some stuff along the way. After the wrapper was added to the formDialogWide, it no longer took up any space in the normalflow, which would shift elements down to make room for it.

The formDialogWide will be one of the documented special cases for this. As it, contrary to the other validation texts, needs to stay in the normal document flow, rather than be taken out. It is possible to make these with absolute position outside the flow like most validation texts, but considering the above, I think letting it stay with static position is the best approach. Even if it was absolutely positioned, It would still have to be a special case due to it's position under the elements, rather than above.

I also noticed that the keyup triggered all of the validation texts to reload in the same form. I was worried this was my work but it turned out to be the way it's always worked. I will therefore not deal with the issue here.

Flipping back and forth between my branch and the original is not very fun, but necessary.

**Comment by a22chrfa** [comment]
Log:
As of now, I think all form in "courseed.php" and "sectioned.php" is covered and is working correctly. Not counting the subpages. Apart from the potential special cases (of which I don't think there's too many) it should now be a matter of finding the validation texts and wrapping them in the dialogwrapper.

Will probably be lots of swapping branches though...

**Comment by a22chrfa** [comment]
Log:
Started with "duggaed.php". These validation texts use entirely different classes and are also positioned with pixels. My intention is to see if these can be restructured to use the same classes as the other validation texts. Seems tricky as one changed made the entire form disappear.

**Comment by a20jonda** [comment]
Hey, I was the one that added .formDialog in 2022 when I was tasked to add error messages to all forms. I believe the form in accessed.php had the original error dialogs that I was tasked to copy to all the others. I believe the difference between inputwrapper and flexwrapper made me create .formDialog.

duggaed.php got the .tooltipDugga class due to it being flexwrapper when I added the errors.

**Comment by a20jonda** [comment]
I believe most of the problems occurred because of the attribute `position:` `relative` and `absolute`.
`relative`  ment that the dialogs where aligned differently depending on the content in the wrapper. `absolute` ment that the dialogs would align themselves only to the wrapper but only if the wrapper had the correct properties.

**Comment by a20jonda** [comment]
These are absolute
![image](https://github.com/HGustavs/LenaSYS/assets/102598258/47f696b7-9c0b-47a6-9dca-e9440223e28c)

If i add "position: relative" to the inputwrapper. They not are aligned with each other albeit -100px due to someone adding that to the .formDialog without checking.
![image](https://github.com/HGustavs/LenaSYS/assets/102598258/7dfecf16-3e4f-4162-9b7b-331ba303f22b)

**Comment by a22chrfa** [comment]
@a20jonda Thanks for the info, it's always good to know the intention behind code. And while you probably had a nice worked out solution, someone will take a shortcut somewhere and the code will degrade over time, which you pointed out. This will likely happen after I'm done as well.

Your first comment is absolutely right. The rules are a bit tricky. And it depends on lot's of thing.

I want the dialogwrapper only containing the input and the validation text, because when the validation text is taken out of the flow, only the input remains. Which makes dialogwrapper be the exact same size and position as the input. Which means the dialog can be anchored to the input. This is how we can make all dialogs be it's own width to the left of the input, and vertically centered. If I use the inputwrapper for this, which contains the entire row of elements, we can't guarantee that the validation text is always perfectly aligned with the perfect space in relation to the input. We can make it for the entire row, since the inputwrapper would be it's parent. But if row width in the form ever changes for example, this might make the spacing between the validation texts and the inputs larger or smaller. Which is where we might start to the "-4 px" and such all over the place.

The relative/absolute is a solution that probably won't fit the inputwrapper, but will fit a wrapper only containing the two mentioned elements. No matter what happens to the code **around** those elements, the validation text will always be positioned the same relative to the input. Which is the goal here. But again, it will likely be misused somewhere and the cycle will start again.

**Comment by a22chrfa** [comment]
After looking some more into duggaed.php I realize that this MIGHT be best to handle in a separate issue. If this is to follow a new generalized standard it might take some work to remake this. It might make this issue unnecessarily long. Some of the elements however, does have the same classes so they are now misaligned. I'll check with our fearless leaders how to best deal with this.

For now, I'm moving onto find other elements that need fixing.

**Comment by a22chrfa** [comment]
Log:
In last commit, I removed inline styling and applied the new wrappers to another form found in sectioned.php.

![image](https://github.com/HGustavs/LenaSYS/assets/128893264/6317e857-ceb1-4c12-9d5d-8a1376181491)

**Comment by a20jonda** [comment]
There is a slight problem with the motd. It has 2 error dialogs. 
<img width="156" alt="image" src="https://github.com/HGustavs/LenaSYS/assets/102598258/7e7a943f-c125-43cc-98ec-7d824e4255a5">
I used formDialogSplitUp and formDialogSplitDown to fit them together. However a better solution could be required.

**Comment by a22chrfa** [comment]
@a20jonda Thanks, will get there!

Log:
In recent commit, added the wrappers to the github form on an item.

![image](https://github.com/HGustavs/LenaSYS/assets/128893264/965ad680-1a6a-4bf1-b2ab-0838df5f911a)

**Comment by a22chrfa** [comment]
From last commit. Fixed message of the day wrappers and removed the specific styling mentioned by @a20jonda .

Result:
![image](https://github.com/HGustavs/LenaSYS/assets/128893264/818e46c1-3658-4a25-b364-37739c52bc24)
![image](https://github.com/HGustavs/LenaSYS/assets/128893264/bff51402-c939-40ea-a38b-91c5e321893c)

**Comment by a22chrfa** [comment]
Log:
I think I've covered most forms now. It's a bit hard to tell since creating new items in courses doesn't work at the moment. I've looked for all lines with the formDialog class through the entire project folder and it seems almost everything is covered.

1. The formDialogWide is a special case. There are a few examples of these I need to check before I can consider them done.
2. The "Edit Dugga" form remains a bit unique form the others. I've looked at it again and tried a few different things, but it seems revising this form takes some work. I still lean towards letting this form be a separate issue where. When the previous bullet point is taken care of, this will be the only remaining form that isn't using our new standardised format.

Will now look at the formDialogWide lines and ensure these work as expected. Then I'm pretty much done and can update the wiki.

**Comment by a22chrfa** [comment]
Double checked that formDialogWide looks as expected. It does. It will as mentioned have to be one of the special cases, but this will be described in the wiki.

Regarding the "Edit dugga". I'll await response from leader and if we got green light then I'll start updating the wiki. Goodnight.

**Comment by a22chrfa** [comment]
Adjusting for reponsiveness seems easy enough. We just have to add width to "formDialogText" on different media widths. But I will leave this for the next issue since there are some more aspects to deal with. I did make on change that will make the responsiveness be smoother: I added flex-start to inputwrapper because if the description belonging to the input took up too much vertical space it caused the dialogwrapper to allocate all vertical space within inputwrapper, which meant that the validation text got misaligned vertically. It won't now.

**Comment by a20jonda** [comment]
[#15938](https://github.com/HGustavs/LenaSYS/pull/15938/files) Heads up that there is appear to be some movement to remove all inline style. I think this have a high chance of conflicting with your work.

**Comment by a22chrfa** [comment]
> [#15938](https://github.com/HGustavs/LenaSYS/pull/15938/files) Heads up that there is appear to be some movement to remove all inline style. I think this have a high chance of conflicting with your work.

Thanks, it does appear to cause some conflicts

**Comment by a22chrfa** [comment]
Log:
Will make a new branch and copy all the changes to avoid the conflicts. Should be relatively quick.
